internal/tracer/exporter.go

```diff
@@ -0,0 +1,146 @@
+package tracer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+
+	exporttrace "go.opentelemetry.io/otel/sdk/export/trace"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+)
+
+// Exporter does 3 things:
+// 1) Accepts spans from OpenTelemetry.
+// 2) Stores spans (for now, in memory)
+// 3) Allows consumers to read spans they might want to send elsewhere
+// Numbers 2 and 3 access the same data, and so it's a concurrency issue.
+type Exporter struct {
+
+	// members for communicating with the loop() goroutine
+
+	// for OpenTelemetry Exporter
+	spanDataCh chan *exporttrace.SpanData
+
+	// for SpanSource
+	readReqCh  chan struct{}
+	readRespCh chan readResp
+}
+
+// SpanSource is the interface for consumers (generally telemetry.Controller)
+type SpanSource interface {
+	// GetOutgoingSpans gives a consumer access to spans they should send
+	// Data may be nil (so the client knows there won't be any data)
+	// The client must send a value over doneCh; True indicates the
+	// SpanSource should remove the data read; false indicates SpanSource should retain the data.
+	GetOutgoingSpans() (data io.Reader, doneCh chan<- bool, err error)
+}
+
+func NewExporter(ctx context.Context) *Exporter {
+	r := &Exporter{
+		spanDataCh: make(chan *exporttrace.SpanData),
+		readReqCh:  make(chan struct{}),
+		readRespCh: make(chan readResp),
+	}
+	go r.loop(ctx)
+	return r
+}
+
+func (e *Exporter) loop(ctx context.Context) {
+	// spans that have come in and are waiting to be read by a consumer
+	var queue []*exporttrace.SpanData
+
+	// what a consumer (generally telemetry.Controller) is reading right now
+	// we keep track of it so that if there's a script error, the data isn't lost forever
+	var beingRead []*exporttrace.SpanData
+	var readDoneCh chan bool
+
+	for {
+		select {
+		// New work coming in
+		case sd := <-e.spanDataCh:
+			// add to the queue
+			queue = appendAndTrim(queue, sd)
+		case <-e.readReqCh:
+			// send the queue to the reader
+			if len(queue) == 0 {
+				// There's nothing pending, so send a nil reader and a channel the client can write to (because it's buffered with length 1)
+				e.readRespCh <- readResp{doneCh: make(chan bool, 1)}
+				break
+			}
+			r, err := e.makeReader(queue)
```

Comment:
not a big deal, but i'm surprised that the serialization logic live inside the collector. Most APIs would push that out to a separate layer. And then a lot of this serialization/error-handling wouldn't block the collector loop

Reply:
Good idea; I'll move it around in a bit.